src/Storages/RabbitMQ/RabbitMQConnection.cpp

```diff
@@ -0,0 +1,88 @@
+#include "RabbitMQConnection.h"
+
+#include <common/logger_useful.h>
+#include <IO/WriteHelpers.h>
+
+
+namespace DB
+{
+
+static const auto CONNECT_SLEEP = 200;
+static const auto RETRIES_MAX = 20;
+
+
+RabbitMQConnection::RabbitMQConnection(const RabbitMQConfiguration & configuration_, Poco::Logger * log_)
+    : configuration(configuration_)
+    , log(log_)
+    , event_handler(loop.getLoop(), log)
+{
+}
+
+String RabbitMQConnection::connectionInfoForLog() const
+{
+    return configuration.host + ':' + toString(configuration.port);
+}
+
+bool RabbitMQConnection::isConnected()
+{
+    std::lock_guard lock(mutex);
+    return event_handler.connectionRunning() && connection->usable();
+}
+
+bool RabbitMQConnection::connect()
+{
+    std::lock_guard lock(mutex);
+    if (configuration.connection_string.empty())
+    {
+        AMQP::Login login(configuration.username, configuration.password);
+        AMQP::Address address(configuration.host, configuration.port, login, configuration.vhost, configuration.secure);
+        connection = std::make_unique<AMQP::TcpConnection>(&event_handler, address);
+    }
+    else
+    {
+        AMQP::Address address(configuration.connection_string);
+        connection = std::make_unique<AMQP::TcpConnection>(&event_handler, address);
+    }
+
+    auto cnt_retries = 0;
+    while (!connection->ready() && cnt_retries++ != RETRIES_MAX)
+    {
+        event_handler.iterateLoop();
+        std::this_thread::sleep_for(std::chrono::milliseconds(CONNECT_SLEEP));
+    }
+    return event_handler.connectionRunning();
+}
+
+bool RabbitMQConnection::reconnect()
```

Comment:
In this method we have three sequential mutex locking/unlocking. Don't we want to make all reconnect logic under one lock?